presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -844,18 +845,23 @@ protected RelationType visitTable(Table table, AnalysisContext context)
             // is this a reference to a WITH query?
             String name = table.getName().getSuffix();
 
-            Query query = context.getNamedQuery(name);
-            if (query != null) {
+            WithQuery withQuery = context.getNamedQuery(name);
+            if (withQuery != null) {
+                Query query = withQuery.getQuery();
                 analysis.registerNamedQuery(table, query);
 
                 // re-alias the fields with the name assigned to the query in the WITH declaration
                 RelationType queryDescriptor = analysis.getOutputDescriptor(query);
-                ImmutableList.Builder<Field> fields = ImmutableList.builder();
-                for (Field field : queryDescriptor.getAllFields()) {
-                    fields.add(Field.newQualified(QualifiedName.of(name), field.getName(), field.getType(), false));
-                }
-
-                RelationType descriptor = new RelationType(fields.build());
+                Iterator<String> columnNames = withQuery.getColumnNames().orElse(emptyList()).iterator();
+                ImmutableList fields = queryDescriptor.getAllFields().stream()
+                        .map(field -> Field.newQualified(
+                                QualifiedName.of(name),
+                                columnNames.hasNext() ? Optional.of(columnNames.next()) : field.getName(),
+                                field.getType(),
+                                false))
+                        .collect(toImmutableList());
+
```

Comment:
this is the cleanest way that I came up with for iterating List and Collection together. I hope we can find a better way